flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -142,14 +150,16 @@ public NetworkBufferPool(
 
 	@Nullable
 	public MemorySegment requestMemorySegment() {
-		return availableMemorySegments.poll();
+		synchronized (availabilityHelper) {
+			return internalRequestMemorySegment();
+		}
 	}
 
 	public void recycle(MemorySegment segment) {
 		// Adds the segment back to the queue, which does not immediately free the memory
 		// however, since this happens when references to the global pool are also released,
 		// making the availableMemorySegments queue and its contained object reclaimable
-		availableMemorySegments.add(checkNotNull(segment));
+		internalRecycle(Collections.singleton(checkNotNull(segment)));
```

Comment:
nit: In order to reuse the common codes we wrap the segment to increase some memory overhead, but I guess this overhead could be ignored.

Reply:
Agree that the overhead could be ignored.